File: backend/app/presentation/schemas/employees.py
"""Pydantic-схемы CRUD сотрудников.

`EmployeeResponse` намеренно не содержит `hashed_password` — даже
маскированный hash наружу не пробрасываем.
"""

from __future__ import annotations

from datetime import time

from pydantic import BaseModel, ConfigDict, EmailStr, Field, SecretStr

from app.domain.employees.entities import Role


class EmployeeCreateRequest(BaseModel):
    """POST /api/v1/employees — admin создаёт сотрудника."""

    model_config = ConfigDict(extra="forbid")

    email: EmailStr
    full_name: str = Field(..., min_length=1, max_length=255)
    role: Role
    initial_password: SecretStr = Field(..., min_length=8, max_length=128)
    schedule_start: time | None = None
    schedule_end: time | None = None


class EmployeeUpdateRequest(BaseModel):
    """PATCH /api/v1/employees/{id} — все поля опциональны."""

    model_config = ConfigDict(extra="forbid")

    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    role: Role | None = None
    schedule_start: time | None = None
    schedule_end: time | None = None
    # Явные флаги «очистить расписание»: иначе нельзя различить
    # «не передал поле» (PATCH-семантика «не менять») и «передал null»
    # («убрать»). Pydantic эти семантики путает на JSON.
    clear_schedule_start: bool = False
    clear_schedule_end: bool = False


class ChangePasswordRequest(BaseModel):
    """POST /api/v1/employees/{id}/password.

    Если вызывает admin для другого пользователя — `old_password`
    не передаётся (admin-reset режим). Если сотрудник меняет свой —
    `old_password` обязателен; endpoint валидирует это правило.
    """

    model_config = ConfigDict(extra="forbid")

    new_password: SecretStr = Field(..., min_length=8, max_length=128)
    old_password: SecretStr | None = None


class EmployeeResponse(BaseModel):
    """Публичная информация о сотруднике (без hashed_password!)."""

    model_config = ConfigDict(extra="forbid")

    id: int
    email: EmailStr
    full_name: str
    role: str
    is_active: bool
    schedule_start: time | None = None
    schedule_end: time | None = None


class EmployeesPageResponse(BaseModel):
    """Страница сотрудников с пагинацией."""

    model_config = ConfigDict(extra="forbid")

    items: list[EmployeeResponse]
    total: int
    limit: int
    offset: int

File: web/src/api/endpoints/employees.ts
import { apiClient } from "../client";
import type {
  EmployeeCreateRequest,
  EmployeeResponse,
  EmployeeUpdateRequest,
  EmployeesPageResponse,
  Role,
  SetPasswordRequest,
} from "../types";

export interface ListEmployeesParams {
  limit?: number;
  offset?: number;
  search?: string;
  role?: Role;
  is_active?: boolean;
}

export const employeesApi = {
  async list(params: ListEmployeesParams = {}): Promise<EmployeesPageResponse> {
    const { data } = await apiClient.get<EmployeesPageResponse>("/v1/employees", { params });
    return data;
  },

  async get(id: number): Promise<EmployeeResponse> {
    const { data } = await apiClient.get<EmployeeResponse>(`/v1/employees/${id}`);
    return data;
  },

  async create(req: EmployeeCreateRequest): Promise<EmployeeResponse> {
    const { data } = await apiClient.post<EmployeeResponse>("/v1/employees", req);
    return data;
  },

  async update(id: number, patch: EmployeeUpdateRequest): Promise<EmployeeResponse> {
    const { data } = await apiClient.patch<EmployeeResponse>(`/v1/employees/${id}`, patch);
    return data;
  },

  async setPassword(id: number, req: SetPasswordRequest): Promise<void> {
    await apiClient.post(`/v1/employees/${id}/password`, req);
  },

  async deactivate(id: number): Promise<EmployeeResponse> {
    const { data } = await apiClient.post<EmployeeResponse>(`/v1/employees/${id}/deactivate`);
    return data;
  },

  async activate(id: number): Promise<EmployeeResponse> {
    const { data } = await apiClient.post<EmployeeResponse>(`/v1/employees/${id}/activate`);
    return data;
  },
};
